src/libcore/array.rs

```diff
@@ -72,215 +72,240 @@ impl TryFromSliceError {
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<'a, 'b, A: Sized, B, const N: usize> PartialEq<[A; N]> for &'b mut [B] where B: PartialEq<A> {
+    #[inline]
+    fn eq(&self, other: &[A; N]) -> bool { self[..] == other[..] }
+    #[inline]
+    fn ne(&self, other: &[A; N]) -> bool { self[..] != other[..] }
+}
+
+// NOTE: some less important impls are omitted to reduce code bloat
+// after const generics came true in 2019, we didn't immediate add these back.
+
+/*
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<'a, 'b, A: Sized, B, const N: usize> PartialEq<&'b [B; N]> for [A; N] where A: PartialEq<B> {
+    #[inline]
+    fn eq(&self, other: &&'b [B; N]) -> bool { self[..] == other[..] }
+    #[inline]
+    fn ne(&self, other: &&'b [B; N]) -> bool { self[..] != other[..] }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<'a, 'b, A: Sized, B, const N: usize> PartialEq<[A; N]> for &'b [B; N] where B: PartialEq<A> {
+    #[inline]
+    fn eq(&self, other: &[A; N]) -> bool { self[..] == other[..] }
+    #[inline]
+    fn ne(&self, other: &[A; N]) -> bool { self[..] != other[..] }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<'a, 'b, A: Sized, B, const N: usize> PartialEq<&'b mut [B; N]> for [A; N]
+    where A: PartialEq<B> {
+    #[inline]
+    fn eq(&self, other: &&'b mut [B; N]) -> bool { self[..] == other[..] }
+    #[inline]
+    fn ne(&self, other: &&'b mut [B; N]) -> bool { self[..] != other[..] }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<'a, 'b, A: Sized, B, const N: usize> PartialEq<[A; N]> for &'b mut [B; N]
+    where B: PartialEq<A> {
+    #[inline]
+    fn eq(&self, other: &[A; N]) -> bool { self[..] == other[..] }
+    #[inline]
+    fn ne(&self, other: &[A; N]) -> bool { self[..] != other[..] }
+}
+
+*/
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<T:Eq, const N: usize> Eq for [T; N] { }
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<T:PartialOrd, const N: usize> PartialOrd for [T; N] {
+    #[inline]
+    fn partial_cmp(&self, other: &[T; N]) -> Option<Ordering> {
+        PartialOrd::partial_cmp(&&self[..], &&other[..])
+    }
+    #[inline]
+    fn lt(&self, other: &[T; N]) -> bool {
+        PartialOrd::lt(&&self[..], &&other[..])
+    }
+    #[inline]
+    fn le(&self, other: &[T; N]) -> bool {
+        PartialOrd::le(&&self[..], &&other[..])
+    }
+    #[inline]
+    fn ge(&self, other: &[T; N]) -> bool {
+        PartialOrd::ge(&&self[..], &&other[..])
+    }
+    #[inline]
+    fn gt(&self, other: &[T; N]) -> bool {
+        PartialOrd::gt(&&self[..], &&other[..])
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<T:Ord, const N: usize> Ord for [T; N] {
+    #[inline]
+    fn cmp(&self, other: &[T; N]) -> Ordering {
+        Ord::cmp(&&self[..], &&other[..])
+    }
+}
+
+#[stable(since = "1.4.0", feature = "array_default")]
+impl<T, const N: usize> Default for [T; N] where T: Default {
```

Comment:
The current `Default` impl for `[T; 0]` doesn't have a `T: Default` bound so this would be a breaking change.

Reply:
In the short term, i think we can put the original definition back. In the long term, how would we solve this? `N > 0` or `T: !Default` or some overlapping specialization?

@Centril

Reply:
If there was a good way to do this we could remove the bounds on the other `[T; 0]` impls as well: #52246.

Reply:
Seems like an obvious candidate for specialization, though unfortunately I'm not aware of any existing efforts towards const-specialization. This is in the "minimal trivially sound" category, though, so again I'd personally be inclined to accept whatever minor hack we need to make it "just work".

Reply:
I'm surprised that the unconstrained `impl Default for [T; 0]` even exists given https://github.com/rust-lang/rust/issues/52246.
Is it some early pre-1.0 addition that was overlooked during stabilization?

Reply:
Introduced in https://github.com/rust-lang/rust/pull/27825 in Rust 1.4 time frame, and the `[T; 0]` special case is never noticed during review :(

Reply:
Couldn't you implement it with a helper trait like this:

```rust
impl<T, const N: usize> Default for [T; N] 
where 
    T: Default 
    Self:DefaultImpl<{N==0}>
{
    fn default()->Self{
        Self::default_impl()
    }
}


trait DefaultImpl<const IS_ZERO:bool>{
    fn default_impl()->Self;
}

impl<T> DefaultImpl<true> for [T;0] {
    fn default_impl()->Self{
        []
    }
}

impl<T,const N:usize> DefaultImpl<false> for [T;0] {
    fn default_impl()->Self{
        [T::default(); N]
    }
}

```

Reply:
Note that serde [has this "[T; 0] doesn't require Trait" optimization too](https://github.com/serde-rs/serde/commit/6388019ad4840a1b5c515ffc353e6a4f2df3adc3). Unless @dtolnay wants to do a breaking change to serde, which I doubt will happen, big array support in native serde is blocked by a language fix for this issue.

Reply:
@est31 you don't think the pattern from @rodrimati1992 would work for Serde?

Reply:
@dtolnay it could possibly work. Not sure yet what const generics will allow in their initial version and what they don't. If I put @rodrimati1992 's code into the playpen I get "identifiers may currently not be used for const generics" errors on the `true` and `false` expressions. If I replace them with `{0u8==0u8}` and `{0u8!=0u8}`, I get tons of ICEs ([link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=83b93a779eb7788a2c2aa73b7b8e2563)).

Reply:
Let's check again after #60742 and #60839.

Reply:
Still ICE's as of `rustc 1.37.0-nightly (3ade426ed 2019-05-30) `. I've filed a bug at #61383

Reply:
I have a question: why does `()` implemented Default, while `[T;0]` does not from the beginning?
They are seemed quite similar for me, if you look `[T; n]` as tuple `(T, T, ... ,T)`. And they have the same memory structure, right?

Reply:
Now there is a compilation error: #61935